pkg/reconciler/pipelinerun/pipelinerun_test.go

```diff
@@ -2569,6 +2597,56 @@ status:
 	}
 }
 
+func TestReconcileWithTimeouts_Tasks(t *testing.T) {
+	// TestReconcileWithTimeouts_Tasks runs "Reconcile" on a PipelineRun with timeouts.tasks configured.
+	// It verifies that reconcile is successful, the pipeline status updated and events generated.
+	ps := []*v1beta1.Pipeline{simpleHelloWorldPipeline}
+	prs := []*v1beta1.PipelineRun{parse.MustParsePipelineRun(t, `
+metadata:
+  name: test-pipeline-run-with-timeout
+  namespace: foo
+spec:
+  pipelineRef:
+    name: test-pipeline
+  serviceAccountName: test-sa
+  timeouts:
+    tasks: 15m
```

Comment:
@lbernick - here's a test verifying that the new `TaskRun` gets a timeout of `timeouts.Tasks` minus the current run time of the `PipelineRun`.

Reply:
sorry, I still think this doesn't work. The pipelinerun started 1 day before ["now"](https://github.com/tektoncd/pipeline/blob/a0c7d31e23f472cca39690410253a0739f3189db/pkg/reconciler/pipelinerun/pipelinerun_test.go#L104-L105), but its tasks timeout is 15 minutes. This means the task has already exceeded its timeout and will be created with a timeout of 1s [here](https://github.com/tektoncd/pipeline/blob/main/pkg/reconciler/pipelinerun/pipelinerun.go#L1154-L1156). (That's a known bug I've been wrestling with.)

I think the test case we actually want is a pipelinerun with some elapsed time having occurred, but still not having passed its "tasks" timeout. Previously, we would have set the taskrun timeout to the tasks timeout minus the elapsed time (covered in the test case below called "taskrun with elapsed time; task.timeout applies"), but it sounds like now, the new approach will be for the pipelinerun to cancel all taskruns when the "tasks" timeout is exceeded.

If it would be helpful, I can create a PR adding some test cases to the PR reconciler with what I have in mind and you can rebase onto that, so it's easier to tell what your PR changes? (I think it will be helpful regardless)

Reply:
Right, that's actually what this test does - the `PipelineRun` started 5 minutes before "now" - `startTime: "2021-12-31T23:55:00Z"`. =)

Reply:
oh oops I didn't look at the start time carefully enough!